projects/plugins/wpcomsh/wpcomsh.php

```diff
@@ -217,26 +217,37 @@ function wpcomsh_jetpack_sso_auth_cookie_expiration( $seconds ) {
 add_filter( 'jetpack_sso_auth_cookie_expiration', 'wpcomsh_jetpack_sso_auth_cookie_expiration' );
 
 /**
- * Determine if users who are already logged in to WordPress.com are automatically logged in to wp-admin.
+ * Determine if users should be enforced to log in with their WP.com account.
+ *
+ * Sites without local users:
+ * - WP.com login, always.
+ *
+ * Sites with local users:
+ * - If user comes from Calypso: WP.com login
+ * - Otherwise: Jetpack SSO login, so they can decide whether to use a WP.com account or a local account.
  */
 function wpcomsh_bypass_jetpack_sso_login() {
-	/**
-	 * Sites with the classic interface:
-	 * - Automatic login if they come from Calypso.
-	 * - Otherwise we display the login form, so they can decide whether to use a WP.com account or a local account.
-	 */
-	if ( 'wp-admin' === get_option( 'wpcom_admin_interface' ) ) {
-		$calypso_domains = array(
-			'https://wordpress.com/',
-			'https://horizon.wordpress.com/',
-			'https://wpcalypso.wordpress.com/',
-			'http://calypso.localhost:3000/',
-			'http://127.0.0.1:41050/', // Desktop App.
-		);
-		return in_array( wp_get_referer(), $calypso_domains, true );
+	$calypso_domains = array(
+		'https://wordpress.com/',
+		'https://horizon.wordpress.com/',
+		'https://wpcalypso.wordpress.com/',
+		'http://calypso.localhost:3000/',
+		'http://127.0.0.1:41050/', // Desktop App.
+	);
+	if ( in_array( wp_get_referer(), $calypso_domains, true ) ) {
+		return true;
+	}
+
+	if ( class_exists( '\Automattic\Jetpack\Connection\Manager' ) ) {
+		$connection_manager = new \Automattic\Jetpack\Connection\Manager( 'jetpack' );
+		$users              = get_users( array( 'fields' => array( 'ID' ) ) );
+		foreach ( $users as $user ) {
+			if ( ! $connection_manager->is_user_connected( $user->ID ) ) {
```

Comment:
How does this perform for large sites? > 10k users?

Reply:
Good question! I have no idea 😅 I'll see if there is an easy way to populate a test site with 10k fake users and check

Reply:
Used FakerPress to create 5k users (not connected to WP.com) on a site:

<img width="1712" alt="Screenshot 2024-09-04 at 11 22 11" src="https://github.com/user-attachments/assets/a9d265e7-e229-4765-bee9-b792e670690c">

It seems that the performance is not largely affected in this scenario.

Before, it took 1.34s to show the Jetpack SSO login. Now, it takes 1.53s

Before | After
--- | ---
<img width="1193" alt="Screenshot 2024-09-04 at 11 24 03" src="https://github.com/user-attachments/assets/94ef7e3d-8a58-42e7-9735-6fdd1abb18e6"> | <img width="1193" alt="Screenshot 2024-09-04 at 11 22 51" src="https://github.com/user-attachments/assets/416db575-78b7-4f41-8ced-6aea4b2aa5d4">

Reply:
Imo, that's just one scenario. If 4.9k users are connected there would be a higher performance penalty. This has an O(n) complexity.